src/EventStore.Common/Utils/CertificateExtensions.cs

```diff
@@ -0,0 +1,141 @@
+				var asnReader = new AsnReader(extension.RawData, AsnEncodingRules.DER).ReadSequence();
+				while (asnReader.HasData) {
+					Asn1Tag tag;
+					try {
+						tag = asnReader.PeekTag();
+					} catch (AsnContentException) {
+						break;
+					}
+
+					switch (tag.TagValue) {
+						case 2: // dNSName [2] IA5String
+							sans.Add((asnReader.ReadCharacterString(UniversalTagNumber.IA5String, tag), CertificateNameType.DnsName));
+							break;
+						case 7: // iPAddress [7] OCTET STRING
+							sans.Add((new IPAddress(asnReader.ReadOctetString(tag)).ToString(), CertificateNameType.IpAddress));
+							break;
+						default:
+							asnReader.ReadEncodedValue();
+							break;
+					}
+				}
+			}
+
+			return sans;
+		}
+
+		public static bool MatchesName(this X509Certificate certificate, string name) {
+			// Implemented based on RFC 6125 (https://datatracker.ietf.org/doc/html/rfc6125) with the following changes:
+			// - Does not support SRV-ID and URI-ID identifier types yet
+			// - Partial wildcard support is not implemented since it has been deprecated in most major browsers
+
+			if (certificate is not X509Certificate2 certificate2) {
+				using var c2 = new X509Certificate2(certificate);
+				certificate2 = c2;
+			}
+
+			var sans = GetSubjectAlternativeNames(certificate2).ToArray();
+			if (sans.Length > 0)
+				return sans.Any(san => MatchesName(san.name, san.type, name));
+
+			var cn = GetCommonName(certificate2);
+			return cn != null && MatchesName(cn, CertificateNameType.DnsName, name);
+		}
+
+		private static string GetCommonName(X509Certificate2 certificate) => certificate.GetNameInfo(X509NameType.SimpleName, false);
+
+		private static bool HasNonAsciiChars(string s) => s.Any(t => t > 127);
+
+		private static bool IsInternationalizedDomainNameLabel(string s) {
+			const string ACEPrefix = "xn--";
+			return HasNonAsciiChars(s) || s.StartsWith(ACEPrefix, StringComparison.OrdinalIgnoreCase);
+		}
+
+		// Based on RFC 952
+		private static bool IsValidDnsNameLabel(string label) =>
+			label.All(x => x is >= '0' and <= '9' or >= 'a' and <= 'z' or >= 'A' and <= 'Z' or '-');
+
+		// Based on RFC 6125 (without support for partial wildcard DNS labels
+		// since it has been deprecated by most major browsers)
+		private static bool IsValidCertificateNameFirstLabel(string label) =>
+			label == "*" || IsValidDnsNameLabel(label);
+
+		private static bool MatchesName(string certName, string certNameType, string name) {
+			const char Wildcard = '*';
+			const char Delimiter = '.';
+
+			if (string.IsNullOrEmpty(certName) ||
+			    string.IsNullOrEmpty(name))
+				return false;
+
+			// if at least one of the names is an IP address, do an exact match
+			if (certNameType == CertificateNameType.IpAddress ||
+			    IPAddress.TryParse(certName, out _) ||
+			    IPAddress.TryParse(name, out _))
+				return name.EqualsOrdinalIgnoreCase(certName);
```

Comment:
We document that it is a requirement that if an IP address is supplied in the certificate then it has to match the machine's IP address regardless of the DNS name provided in the SAN.

Reply:
@hayley-jean i didn't follow - actually, we do check for all SANs here:

https://github.com/EventStore/EventStore/blob/803c2b55ce8834918aaaafff090099233f75555a/src/EventStore.Common/Utils/CertificateExtensions.cs#L71

Reply:
but if either the san or the target host is an ip address then we do an exact check (since a hostname cannot contain only digits according to RFCs 952 & 1123)

Reply:
Thanks, I didn't notice that. Then yeah, all we would need to do is add the fact that you can have wildcard certs with dns discovery to the documentation and leave it at that.